flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunner.java

```diff
@@ -261,17 +260,34 @@ public void grantLeadership(UUID leaderSessionID) {
 
     @GuardedBy("lock")
     private void startJobMasterServiceProcessAsync(UUID leaderSessionId) {
-        sequentialOperation =
-                sequentialOperation.thenRun(
-                        () ->
-                                runIfValidLeader(
-                                        leaderSessionId,
-                                        ThrowingRunnable.unchecked(
-                                                () ->
-                                                        verifyJobSchedulingStatusAndCreateJobMasterServiceProcess(
-                                                                leaderSessionId)),
-                                        "verify job scheduling status and create JobMasterServiceProcess"));
-
+        boolean isValid;
+        synchronized (lock) {
+            isValid = isValidLeader(leaderSessionId);
```

Comment:
That goes into the right direction. :+1:  The leader validity should be checked when calling the methods (i.e. when checking the `jobResultStore` in the `thenCompose` and the handling of the response in the `handle` method).

Reply:
We still need to use the `runIfValidLeader`. But we need to adapt it in a way that we also allow the returning of a value (for the `hasJobResultEntryAsync` return).